README.md
Laravel Data Visualizer
---
With Laravel Data Visualizer you can render data from a database as svg (or save as png).
One advantage is a better view of the stored data than a normal COUNT usage.

## Installing

```shell
$ composer require dobron/data-grid-visualizer
```

## Usage
### Model

```php
echo Patient::query()->dataGridVisualizer([
    'visualize' => [
        'birth_year',
        function ($birthYear) {
            if ($birthYear < 1930) {
                return '#0e4429';
            }

            if ($birthYear < 1960) {
                return '#006d32';
            }

            if ($birthYear < 1990) {
                return '#26a641';
            }

            return '#39d353';
        }
    ],
])
```

### Without Model
```php
DB::table('patients')->whereIn('birth_year', [1996, 2000])->dataGridVisualizer([
    'visualize' => [
        'birth_year',
        [
            1996 => '#800000'
            2000 => '#ffc800'
        ],
    ],
]);
```

### Render as PNG
```php
$query->dataGridVisualizer([...])->image();
```

### Save as PNG
```php
$query->dataGridVisualizer([...])->image('graph.png');
```

## Options

| Option  | Description                                                        | Default                   |
| ------- |------------------------------------------------------------------- | ------------------------- |
| defaultColor | Default grid color if none is defined.                        | #2d333b                   |
| gap          | Gap size in px.                                               | 13                        |
| size         | Size of grid in px.                                           | 10                        |
| radius       | Radius of the grid.                                           | 2                         |
| outline      | Outline color of the grid (optional).                         | rgba(255, 255, 255, 0.05) |
| perspective  | Data view perspective (horizontal/vertical).                  | horizontal                |
| visualize    | Visualization conditions as array [column name, condition].   | *none*                    |

## Examples

<img src="./images/visualization1.png">
<img src="./images/visualization2.png">

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.


## Testing

``` bash
$ composer test
```


## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.


## Credits

- [Richard Dobroň][link-author]


## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[link-author]: https://github.com/richardDobron


CHANGELOG.md
# Changelog

All notable changes to `data-grid-visualizer` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## v1.0.0 - 2021-12-21

### Added
- Added Laravel query macro to generate SVG visualizations.


tests/Patient.php
<?php

namespace tests;

use Illuminate\Database\Eloquent\Model;

class Patient extends Model
{
    protected $fillable = ['birth_year'];
}


tests/Product.php
<?php

namespace tests;

use Illuminate\Database\Eloquent\Model;

class Product extends Model
{
    protected $fillable = ['visible'];
}


tests/TestCase.php
<?php

namespace tests;

use dobron\DataGridVisualizer\DataGridVisualizerServiceProvider;

abstract class TestCase extends \Orchestra\Testbench\TestCase
{
    /**
     * Load package service provider.
     *
     * @param \Illuminate\Foundation\Application $app
     *
     * @return array
     */
    protected function getPackageProviders($app)
    {
        return [DataGridVisualizerServiceProvider::class];
    }

    /**
     * Define environment setup.
     *
     * @param \Illuminate\Foundation\Application $app
     */
    protected function getEnvironmentSetUp($app)
    {
        // Setup default database to use sqlite :memory:
        $app['config']->set('database.default', 'testing');
        $app['config']->set('database.connections.testing', [
            'driver' => 'sqlite',
            'database' => ':memory:',
            'prefix' => '',
        ]);
    }

    protected function setUp(): void
    {
        parent::setUp();

        $this->loadMigrationsFrom(__DIR__.'/migrations');
    }
}


src/Svg.php
<?php

namespace dobron\DataGridVisualizer;

use Imagick;

class Svg
{
    /**
     * @var string
     */
    protected $content;
    /**
     * @var array
     */
    protected $attributes;

    public function __construct(string $content, array $attributes = [])
    {
        $this->content = $content;
        $this->attributes = $attributes;
    }

    public function __toString()
    {
        $svg = "<svg xmlns=\"http://www.w3.org/2000/svg\"";
        foreach ($this->attributes as $attribute => $value) {
            if (!$value) {
                continue;
            }

            $svg .= " " . $attribute . "=\"" . htmlspecialchars($value, ENT_QUOTES) . "\"";
        }

        $svg .= ">" . $this->content . "</svg>";
        return $svg;
    }


    public function image(?string $filename = null)
    {
        $svg = '<?xml version="1.0" encoding="UTF-8" standalone="no"?>' . $this;

        $image = new Imagick();
        $image->setImageFormat("png32");
        $image->readImageBlob($svg);

        if ($filename) {
            return $image->writeImage($filename);
        }

        header("Content-type: image/png");

        return $image->getImageBlob();
    }
}


.php-cs-fixer.php
<?php

return (new PhpCsFixer\Config())
    ->setRules([
        '@PSR12' => true,
        'binary_operator_spaces' => true,
        'blank_line_after_opening_tag' => true,
        'compact_nullable_typehint' => true,
        'declare_equal_normalize' => true,
        'lowercase_cast' => true,
        'lowercase_static_reference' => true,
        'new_with_braces' => true,
        'no_blank_lines_after_class_opening' => true,
        'no_leading_import_slash' => true,
        'no_whitespace_in_blank_line' => true,
        'ordered_class_elements' => [
            'order' => [
                'use_trait',
            ],
        ],
        'ordered_imports' => [
            'imports_order' => [
                'class',
                'function',
                'const',
            ],
            'sort_algorithm' => 'none',
        ],
        'return_type_declaration' => true,
        'short_scalar_cast' => true,
        'single_blank_line_before_namespace' => true,
        'single_trait_insert_per_statement' => true,
        'ternary_operator_spaces' => true,
        'unary_operator_spaces' => true,
        'visibility_required' => [
            'elements' => [
                'const',
                'method',
                'property',
            ],
        ],
    ])
    ->setFinder(
        PhpCsFixer\Finder::create()
            ->exclude('vendor')
            ->in([__DIR__.'/src/', __DIR__.'/tests/'])
    )
    ;

src/DataGridVisualizer.php
<?php

namespace dobron\DataGridVisualizer;

use dobron\DataGridVisualizer\Support\Config;
use Illuminate\Support\Collection;

class DataGridVisualizer
{
    /**
     * @var Collection
     */
    private $data;
    /**
     * @var Config
     */
    private $config;

    public function __construct(Collection $data, array $config)
    {
        $this->data = $data;
        $this->config = new Config($config);
    }

    public function prepareMatrixData(): Matrix
    {
        $total = $this->data->count();
        $squareSide = ceil(sqrt($total));
        $gridSize = pow($squareSide, 2);
        $missing = [];

        if ($gridSize > $total) {
            $missing = array_fill($total, $gridSize - $total, $this->config->getDefaultColor());
        }

        [$key, $visualizer] = $this->config->getVisualize();

        return new Matrix($this->data->map(function ($model) use ($key, $visualizer) {
            $value = $model->{$key};

            if (is_callable($visualizer)) {
                $color = $visualizer($value);
            } elseif (is_array($visualizer)) {
                $color = $visualizer[$value] ?? null;
            } elseif (is_string($visualizer)) {
                $color = $visualizer;
            }

            return $color ?? $this->config->getDefaultColor();
        })->concat($missing)->chunk($squareSide)->map(function ($item) {
            return $item->values();
        })->toArray());
    }

    public function renderSvg(Matrix $matrix): Svg
    {
        $horizontal = $this->config->getPerspective() === Config::PERSPECTIVE_HORIZONTAL;
        $radius = $this->config->getRadius();
        $size = $this->config->getSize();
        $gap = $this->config->getGap();
        $w = $matrix->getColumns();
        $h = $matrix->getRows();

        $args = ["viewBox" => "0 0 " . ($w * $gap) . " " . ($h * $gap)];
        $svg = '';

        for ($y = 0; $y < $h; $y++) {
            $svg .= '<g transform="translate(' . ($y * $gap) . ', 0)">';

            for ($x = 0; $x < $w; $x++) {
                $svg .= '<rect';
                $attributes = [
                    'width' => $size,
                    'height' => $size,
                    'x' => $h - $y,
                    'y' => $x * $gap,
                    'rx' => $radius,
                    'ry' => $radius,
                    'fill' => $horizontal
                        ? $matrix->getValue($x, $y)
                        : $matrix->getValue($y, $x),
                    'style' => (
                        ($outline = $this->config->getOutline())
                        ? "outline: 1px solid $outline;outline-offset: -1px"
                        : null
                    ),
                ];

                foreach ($attributes as $attribute => $value) {
                    if (!$value) {
                        continue;
                    }

                    $svg .= " " . $attribute . "=\"" . htmlspecialchars($value, ENT_QUOTES) . "\"";
                }

                $svg .= '></rect>';
            }

            $svg .= '</g>';
        }

        return new Svg($svg, $args);
    }

    /**
     * @return string|Svg
     */
    public function render()
    {
        return $this->renderSvg($this->prepareMatrixData());
    }
}


src/Support/Config.php
<?php

namespace dobron\DataGridVisualizer\Support;

class Config
{
    /**
     * @var array
     */
    protected $options = [
        'defaultColor' => '#2d333b',
        'gap' => 13,
        'size' => 10,
        'radius' => 2,
        'outline' => 'rgba(255, 255, 255, 0.05)',
        'perspective' => self::PERSPECTIVE_HORIZONTAL,
        'visualize' => [],
    ];

    public const PERSPECTIVE_HORIZONTAL = 'horizontal';
    public const PERSPECTIVE_VERTICAL = 'vertical';

    /**
     * @param array $options
     */
    public function __construct(array $options = [])
    {
        $this->options = array_merge($this->options, $options);
    }

    /**
     * @return string
     */
    public function getDefaultColor(): string
    {
        return $this->options['defaultColor'];
    }

    /**
     * @return int
     */
    public function getGap(): int
    {
        return $this->options['gap'];
    }

    /**
     * @return int
     */
    public function getSize(): int
    {
        return $this->options['size'];
    }

    /**
     * @return int
     */
    public function getRadius(): int
    {
        return $this->options['radius'];
    }

    /**
     * @return null|string
     */
    public function getOutline(): ?string
    {
        return $this->options['outline'];
    }

    /**
     * @return string
     */
    public function getPerspective(): string
    {
        return $this->options['perspective'];
    }

    /**
     * @return array
     */
    public function getVisualize(): array
    {
        return $this->options['visualize'];
    }

    /**
     * @return array
     */
    public function toArray(): array
    {
        return $this->options;
    }

    /**
     * @param string $key
     * @param mixed  $value
     *
     * @return $this
     */
    public function setOption(string $key, $value): self
    {
        $this->options[$key] = $value;

        return $this;
    }

    /**
     * @param string $key
     * @param mixed  $default
     *
     * @return mixed
     */
    public function getOption(string $key, $default = null)
    {
        return $this->options[$key] ?? $default;
    }

    /**
     * @param array $options
     *
     * @return $this
     */
    public function mergeOptions(array $options): self
    {
        $this->options = array_merge($this->options, $options);

        return $this;
    }

    /**
     * @param array $options
     *
     * @return $this
     */
    public function setOptions(array $options): self
    {
        $this->options = $options;

        return $this;
    }

    /**
     * @return array
     */
    public function getOptions(): array
    {
        return $this->options;
    }
}
